includes/validation/class-amp-validated-url-post-type.php

```diff
@@ -931,6 +938,45 @@ public static function store_validation_errors( $validation_errors, $url, $args
 		return $post_id;
 	}
 
+	/**
+	 * Delete batch of stylesheets postmeta.
+	 *
+	 * Given that parsed CSS can be quite large (250KB+) and is not de-duplicated across each validated URL, it is important
+	 * to not store the stylesheet data indefinitely in order to not excessively bloat the database. The reality is that
+	 * keeping around the parsed stylesheet data is of little value given that it will quickly go stale as themes and
+	 * plugins are updated.
+	 *
+	 * @since 2.0
+	 *
+	 * @param int          $count  Count of batch size to delete. Accepts strtotime()-compatible string, or array of 'year', 'month', 'day' values.
+	 * @param string|array $before Date before which to find amp_validated_url posts to delete.
+	 * @return int Count of postmeta that were deleted..
+	 */
+	public static function delete_stylesheets_postmeta_batch( $count, $before ) {
+		$deleted = 0;
+		$query   = new WP_Query(
+			[
+				'post_type'      => self::POST_TYPE_SLUG,
+				'meta_key'       => self::STYLESHEETS_POST_META_KEY,
+				'date_query'     => [
+					[
+						'before' => $before,
+					],
+				],
+				'fields'         => 'ids',
+				'orderby'        => 'date',
+				'order'          => 'ASC',
+				'posts_per_page' => $count,
+			]
+		);
+		foreach ( $query->get_posts() as $post_id ) {
+			if ( delete_post_meta( $post_id, self::STYLESHEETS_POST_META_KEY ) ) {
+				$deleted++;
+			}
+		}
```

Comment:
Given the accidental deletion of blocks in 1.5.0, let's make super super sure that this is only deleting the expected postmeta.

Reply:
Checked the following:
- `post_type` syntax looks correct and restricts to validated URL posts only
- `meta_key` syntax looks correct and restricts to only validated URL posts with existing stylesheet postmeta
- `date_query` syntax looks correct and restricts to validated URL posts that are older than 1 week.
- `fields` makes sure we only use IDs to avoid overhead.
- `orderby` & `order` syntax looks correct and ensures we delete the `$count` oldest ones.
- `posts_per_page` syntax looks correct and makes sure we limit to the `$count` oldest posts.

Reply:
We could consider adding safeguards to the `$count` and `$before` arguments, to make sure you can't call them with bad arguments. However, we're currently hard-coding this in the background task, so I don't see how that could go wrong...

Reply:
In terms of performance, what we are doing here is the N+1 problem: We do 1 query to see what to delete, and then we iterate over IDs to do N queries with N cache updates to delete the post meta.

So this would do 101 queries for a complete cleanup iteration per cron processing.

Performance-wise, it would be much better to do 1 single `DELETE` query instead, but it would require a hard-coded SQL query instead of using WP functions...

Reply:
Yes, that's true. However, that will not give an opportunity for object caches to be flushed. Notice that core uses `wp_delete_post()` in `wp_delete_auto_drafts()` rather than a SQL `DELETE`. 

It does not do the same in `_wp_batch_update_comment_type()` when it does a SQL `UPDATE`, but I'd consider that to be a bug in 5.5 because the updated `comment_type` will not be reflected when `wp_cache_get( $comment_id, 'comment' )` is called.

Reply:
Raised my concern about `_wp_batch_update_comment_type()`: https://core.trac.wordpress.org/ticket/49236#comment:24

Reply:
In 66abca1 I've added added explicit tests to ensure that all untargeted postmeta remains intact.